### PR TITLE
scheduler, evict-slow-store allow draining multiple store from AZ

### LINKED ISSUE #10742: evict-slow-store: allow draining multiple nodes from the same availability zone
### Background

The evict-slow-store scheduler transfers leaders away from stores whose disk slow-score exceeds a threshold. It has a deliberate safety guard: if more than one store is slow at the same time, it does nothing.

This is by design — multiple stores degrading simultaneously is a signal of broader cluster instability. Draining leaders in an already-unstable cluster risks making things worse, so the conservative choice is to hold.

### Problem

In cloud environments, an Availability Zone is a single failure domain. A bad deployment, a datacenter power event, or a batch of faulty hardware can cause every node in one AZ to degrade together — while nodes in other AZs remain perfectly healthy. In this situation, all the slow stores share the same AZ label and the cross-zone topology is intact.

The current guard cannot distinguish this case from genuine cluster-wide instability. It sees multiple slow stores and stops, leaving the degraded AZ serving leaders it cannot handle efficiently, when the right action is to drain the whole AZ and let the healthy zones take over.

### Proposal

When all slow stores belong to the same isolation zone, treat the event as an AZ-level failure and drain them all — subject to guards that preserve the original protection against cluster-wide instability:

1. Uniform placement-rule isolation.
Read IsolationLevel from every active placement rule via RuleManager.GetAllRules(). Group eviction only activates when every rule enforces the same non-empty isolation level. If any rule differs or is empty, fall back to the original do-nothing behavior. This ensures the zone concept is actually encoded in the cluster's placement rules, not assumed.

2. All slow stores share one AZ.
Every slow store must carry the same value for the isolation label. If slow stores span two zones, do nothing — this is the genuine multi-domain instability the original guard was designed for.

3. Enough healthy zones remain.
Only drain an AZ if at least 2 other distinct zones still have a healthy (serving, non-slow, leader-transfer-allowed) store. This prevents draining when the topology is too degraded to absorb the load shift.

4. Freeze if a second zone goes slow.
While draining a zone, if any slow score appears in a store outside that zone, stop adding new stores to the eviction set. The situation has changed from "one bad AZ" to "multiple failing domains" — revert to conservative behavior.

5. Group recovery.
Release the drained AZ only once every evicted store has recovered for the configured recovery-duration. The group is treated as a unit: one flapping node does not block recovery of the others, but the whole group must be healthy before leaders flow back.

Relationship to existing slow-store mechanisms

PD has two other mechanisms that partially overlap with this change, and both are intentionally left untouched.

evict-slow-trend detects performance degradation by trend (latency increase, QPS drop) rather than absolute score, and already has a SlowStoreEvictingAffectedStoreRatioThreshold guard (default 0.3) that caps how much of the cluster can be evicted at once. The multi-store guard in that scheduler works differently and is not part of this change.

Network slow store (the secondary path inside evict-slow-store) detects pairwise network degradation between TiKV nodes. It currently caps eviction at one store at a time (defaultMaxNetworkSlowStore = 1) through a separate mechanism: a two-phase pause→evict flow with per-store timers. A zone-level network incident (e.g. a bad switch) could in principle benefit from the same AZ-aware grouping, but the detection logic is more complex and the risk of false positives is higher — that extension is left for a follow-up.

In short

The original guard was designed to protect against an unstable cluster. A full-AZ degradation in a cloud environment is a different event — a known failure domain, not a symptom of cluster-wide problems. This change lets PD tell the difference.

### PR DESCRIPTION
**Pull request by Tema**
### What problem does this PR solve?

Issue Number: Close #10742

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

Use chaos mesh to emulate slow disk
```
apiVersion: chaos-mesh.org/v1alpha1
kind: IOChaos
metadata:
  name: tikv-0
  namespace: chaos-mesh
spec:
  action: latency
  mode: one
  selector:
    namespaces:
      - tidb-dev-test
    labelSelectors:
      'app.kubernetes.io/component': 'tikv'
    fieldSelectors:
      'metadata.name': 'dev-test-tikv-0'
  volumePath: /var/lib/raft
  delay: 100ms
  percent: 100
  duration: '900s'
```

Verify default behavior:
* only first slow node gets drained even in the same AZ
image

Enabling the feature:
* pd-ctl scheduler config evict-slow-store-scheduler set group-eviction-labels "zone"
* Inject disk delay to first node in zone A
* Observe leaders started to drop
* Inject disk delay to a node in zone B
* No additional leaders dropped
* Inject disk delay to a second node in zone A
* No additional leaders dropped
* Remove disk delay from the node in zone B
* Leaders are dropping for the second node in zone A
image

- No code

Code changes

- Has the configuration change
- Has HTTP APIs changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

## Summary by CodeRabbit

* **New Features**
  * Grouped multi-store disk slow-store eviction via configurable eviction labels (comma-separated input) for coordinated domain drains; scheduler now reports multi-store eviction state and blocks scheduling while any disk eviction is active.

* **Bug Fixes / Reliability**
  * Safer persistence and recovery for eviction sets with robust add/remove/clear semantics and improved lifecycle handling.

* **Tests**
  * Expanded unit and end-to-end tests for group eviction, reconciliation, recovery, and edge cases.

### CODE REVIEW

**Comment by Tema** [comment]
```
--- FAIL: TestKeyspaceGroupTestSuite (89.36s)
              --- PASS: TestKeyspaceGroupTestSuite/TestAllocNodes (3.08s)
              --- PASS: TestKeyspaceGroupTestSuite/TestAllocNodesUpdate (1.92s)
              --- PASS: TestKeyspaceGroupTestSuite/TestAllocOneNode (3.65s)
              --- PASS: TestKeyspaceGroupTestSuite/TestAllocReplica (3.41s)
              --- PASS: TestKeyspaceGroupTestSuite/TestDefaultKeyspaceGroup (3.70s)
              --- PASS: TestKeyspaceGroupTestSuite/TestSetNodes (1.55s)
              --- FAIL: TestKeyspaceGroupTestSuite/TestUpdateMemberWhenRecovery (72.06s)
```

File: pkg/schedule/schedulers/evict_slow_store.go [hunk 1]
```diff
@@ -764,75 +826,249 @@ func filterNetworkSlowScores(
...
+		return "", false
+	}
+	level := rules[0].IsolationLevel
+	for _, r := range rules[1:] {
+		if r.IsolationLevel != level {
+			return "", false
+		}
+	}
+	return level, true
+}
```
Comment (LykxSassinator) [inline pkg/schedule/schedulers/evict_slow_store.go]: It is not recommended to use IsolationLevel as a target filter label.

A more appropriate approach is to add a manual configuration that determines whether the slow stores belong to the same availability zone (AZ). And the comparison logic should consider that the configuration value could be a multi-level labels, i.e. `zone` or `zone, rack`.

For example, we could name the configuration `evicting-label` and set it as follows:
- `evicting-label: zone`: Multi-node leader eviction is triggered only when the slow stores share the same zone label.
- `evicting-label: zone, rack`: Multi-node leader eviction is triggered only when the slow stores share both the same zone and rack labels.
- Empty (default): The default evicting strategy applies, which falls back to the original design — evicting a single slow node.

The comparison logic can be referenced from:
https://github.com/tikv/pd/blob/51b1226c301ef160bf0b6ba854d0e5f48d107563/pkg/core/store.go#L691-L703.


Rest LGTM.

Comment (Tema) [reply pkg/schedule/schedulers/evict_slow_store.go]: >> It is not recommended to use IsolationLevel as a target filter label.

I'm curious why not? Just to make it configurable differently? I feel if it does not match isolation expectations, it can lead to issues.

Anyway, I don't have a strong opinion on that and like an option to turn it on/off, so I've added `group-eviction-labels` into PD scheduler config as you recommended.  @LykxSassinator PTAL

Comment (LykxSassinator) [reply pkg/schedule/schedulers/evict_slow_store.go]: Correct me if I'm wrong. /cc @rleungx 

FYI, `isolation-level` is used to determine the HA level for the entire TiKV cluster, not to determine the AZ of individual TiKV nodes.

- By default, this value is empty. This means that regardless of the AZs in which TiKV nodes are deployed, if some TiKV nodes go down (marked as the `Down` state), PD will add replicas on other alive TiKV nodes to ensure overall HA stability — irrespective of the AZs of the TiKV nodes.
- If `isolation-level` is set to `zone` in a three‑AZ deployment, and one AZ goes down, PD will **not** trigger operations to add replicas in other AZs, because the isolation level is constrained to `zone`.

In other words, `isolation-level` is **not** the actual tag that reflects which AZ a TiKV node is deployed in. In fact, `LocationLabel` is used to mark the deployment tag that indicates the location of a given TiKV node.

Regarding my earlier suggestion:

> "A more appropriate approach is to add a manual configuration ..."

On one hand, we want to introduce an extra, clear configuration to explicitly enable or disable this newly introduced feature. On the other hand, by allowing users to manually set this configuration based on their requests, users can control the evicting strategy themselves if multiple TiKV nodes exist, exactly as they expect.

**Review by LykxSassinator — APPROVED** [review]

File: pkg/schedule/schedulers/evict_slow_store.go [hunk 2]
```diff
@@ -764,75 +867,261 @@ func filterNetworkSlowScores(
...
+
+// hasEnoughHealthyDomains reports whether at least minRemainingHealthyDomains distinct
+// failure domains (by all configured labels) other than the drained store's domain still
+// have a store able to host leaders. Domains are deduplicated with CompareLocation so the
+// counting stays consistent with the grouping decision.
+func hasEnoughHealthyDomains(cluster sche.SchedulerCluster, labels []string, drained *core.StoreInfo) bool {
+	reps := make([]*core.StoreInfo, 0, minRemainingHealthyDomains)
+	for _, store := range cluster.GetStores() {
+		if !store.IsUp() || store.IsSlow() || store.EvictedAsSlowStore() || !store.AllowLeaderTransferIn() {
```
Comment (rleungx) [inline pkg/schedule/schedulers/evict_slow_store.go:923]: Can we use the same target-store checks as leader transfer here? This predicate can count stores that `StoreStateFilter{TransferLeader: true}` later rejects (down, disconnected, busy, reject-leader), so group eviction may start without two real healthy target domains.

**Review by okJiang — COMMENTED** [review]
Apart from existing comments, lgtm

File: pkg/schedule/schedulers/evict_slow_store.go (hunk 2 above)
Comment (LykxSassinator) [reply pkg/schedule/schedulers/evict_slow_store.go:923]: FYI, you can use a helper function to take the same filtering condition as "leader transfer" does, for example:

```go

func eligibleHealthyLeaderTargets(cluster sche.SchedulerCluster, stores []*core.StoreInfo) []*core.StoreInfo {
    candidates := filter.NewCandidates(stores).
        FilterTarget(
            cluster.GetSchedulerConfig(),
            nil,
            nil,
            &filter.StoreStateFilter{
                ActionScope:  types.EvictSlowStoreScheduler.String(),
                TransferLeader: true,
                OperatorLevel:  constant.Urgent,
            },
        ).
        PickAll()

    healthy := candidates[:0]
    // Filter out already slow nodes
    for _, store := range candidates {
        if store.IsSlow() {
            continue
        }
        healthy = append(healthy, store)
    }
    return healthy
}


```

File: pkg/schedule/schedulers/evict_slow_store.go (hunk 1 above)
Comment (LykxSassinator) [reply pkg/schedule/schedulers/evict_slow_store.go]: After further discussion, I think we should keep this PR simple and revert to the previous isolation-based direction.

The current `group-eviction-labels` config feels unnecessary here. It is not well aligned with the semantics of `IsolationLevel`, and it adds extra config, validation, and operational cost for what should be a focused correctness fix.

But the old implementation still needs one fix: it treated `IsolationLevel` as a single label key, while PD's actual isolation semantics are based on the `LocationLabels` prefix up to and including `IsolationLevel`.

So for this PR, I suggest:
- revert to the previous isolation-based design;
- fix the implementation to compare the `LocationLabels` prefix up to `IsolationLevel`, instead of only `store.GetLabelValue(IsolationLevel)`;
- keep the original conservative rule that grouping is allowed only when all placement rules share the same non-empty `IsolationLevel`;
- add only a simple boolean on/off switch so this feature can be explicitly disabled.

This keeps the change aligned with PD's existing isolation model, fixes the real bug, and avoids turning this PR into a broader config-policy discussion.

@Tema PTAL

Comment (LykxSassinator) [reply pkg/schedule/schedulers/evict_slow_store.go]: @HaoW30 PTAL.
